File: solution/160/answer.cpp
#include <stdio.h>

struct ListNode{
    int val;
    ListNode *next;
    ListNode(int x):val(x), next(NULL){}
};

void printAll(ListNode *head, const char *name){
    printf("%s:", name);
    if(!head){
        printf("NULL\n");
        return;
    }
    while(head){
        printf(" %d", head->val);
        head = head->next;
    }
    printf("\n");
}

int getListNodeLen(ListNode *head){
    int len = 0;
    while(head){
        len++;
        head = head->next;
    };
    return len;
}

ListNode* forward_long_ListNode(int longlen, int shortlen, ListNode *head){
    int subab = longlen - shortlen;
    while(head && subab){
        head = head->next;
        subab--;
    };
    return head;
}

class Solution{
public:
    ListNode* getIntersectionNode(ListNode* headA, ListNode* headB){
        int lenA = getListNodeLen(headA);
        int lenB = getListNodeLen(headB);
        if(lenA>lenB){
            headA = forward_long_ListNode(lenA, lenB, headA);
        }else{
            headB = forward_long_ListNode(lenB, lenA, headB);
        }
        while(headA&&headB){
            if(headA==headB){
                return headA;
            }
            headA = headA->next;
            headB = headB->next;
        }
        return NULL;
    }
};

int main(){
    ListNode a(1);
    ListNode b(2);
    ListNode c(3);
    ListNode d(4);
    ListNode e(5);
    ListNode f(6);
    ListNode g(7);
    a.next = &b;
    b.next = &c;
    c.next = &d;
    e.next = &f;
    f.next = &g;
    g.next = &c;
    ListNode *headA = &a;
    ListNode *headB = &e;
    ListNode *result = NULL;
    Solution s;
    printAll(headA, "headA");
    printAll(headB, "headB");
    result = s.getIntersectionNode(headA, headB);
    printAll(result, "result");
}

File: solution/160/answer.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

    def printval(self):
        print self.val

def getListNodelen(head):
    len = 0
    while head:
        len += 1
        head = head.next
    return len

def forward_long_list(long, short, head):
    sublen = long - short
    while head and sublen:
        head = head.next
        sublen -= 1
    return head

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        lenA = getListNodelen(headA)
        lenB = getListNodelen(headB)
        if lenA > lenB:
            headA = forward_long_list(lenA, lenB, headA)
        else:
            headB = forward_long_list(lenB, lenA, headB)

        while headA and headB:
            if headA == headB:
                return headA
            headA = headA.next
            headB = headB.next
        return None

if __name__ == "__main__":
    a = ListNode(1)
    b = ListNode(2)
    c = ListNode(3)
    d = ListNode(4)
    e = ListNode(5)
    f = ListNode(6)
    g = ListNode(7)
    a.next = b
    b.next = c
    c.next = e
    f.next = g
    g.next = c
    headA = a
    headB = f
    s = Solution()
    result = s.getIntersectionNode(headA, headB)
    result.printval()